### ISSUE TITLE
Unable to start the C/C++ language server. IntelliSense features will be disabled. Error: TypeError: Cannot read property 'setDidOpenTime' of undefined

### ISSUE THREAD

**Issue by gunarkroeger** [microsoft/vscode-cpptools#6981]
**Type: LanguageService**

**Describe the bug**
- OS and Version: Windows 10
- VS Code Version: 1.53.2
- C/C++ Extension Version:
- Other extensions you installed (and if the issue persists after disabling them):
- Does this issue involve using SSH remote to run the extension on a remote machine?:
- A clear and concise description of what the bug is, including information about the workspace (i.e. is the workspace a single project or multiple projects, size of the project, etc).

**Steps to reproduce**

**Expected behavior**

<details>
  <summary><strong>Logs</strong></summary>

```
![Screenshot 2021-02-17 162812](https://user-images.githubusercontent.com/22892870/108235343-a4fe3900-713d-11eb-8c3a-3931857483d3.png)

[Error - 16:25:25] Server initialization failed.
TypeError: Cannot read property 'setDidOpenTime' of undefined
	at Object.didOpen (c:\Users\gunar\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:60328:52)
	at c:\Users\gunar\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:8696:32
	at Array.forEach (<anonymous>)
	at DidOpenTextDocumentFeature.register (c:\Users\gunar\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:8685:42)
	at DidOpenTextDocumentFeature.initialize (c:\Users\gunar\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:8676:18)
	at LanguageClient.initializeFeatures (c:\Users\gunar\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:10684:21)
	at c:\Users\gunar\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:10389:18
```
</details>
I opened a cpp project today and was greeted with these errors. Intellisense not working now. I don't know what caused it.
**Screenshots**

![Screenshot 2021-02-17 162812](https://user-images.githubusercontent.com/22892870/108235674-f5759680-713d-11eb-8113-755a50f8b224.png)

**Additional context**

**Comment by gunarkroeger** [comment]
Also, building and uploading is not working with the Platform.io extension anymore...

**Comment by gunarkroeger** [comment]
Reverting C/C++ Extension to 1.2.0 fixes the issue

**Comment by Colengms** [comment]
Hi @gunarkroeger .  We've occasionally been seeing an issue with VS Code failing to fully install our extension.  (i.e. https://github.com/microsoft/vscode-cpptools/issues/6936 )  Could you try deleting `c:\Users\gunar\.vscode\extensions\ms-vscode.cpptools-1.2.1` and reinstalling the extension?

If that doesn't address the issue, could you enable logging using the setting `"C_Cpp.loggingLevel": "Debug"` and provide the output of the C/C++ output view leading up to repro?  The symptom doesn't make a lot of sense to me, and doesn't occur for everyone.  I'm curious if there is another issue logged prior to the symptom.

**Comment by gunarkroeger** [comment]
Hi @Colengms,
Deleting the folder and reinstalling version 1.2.1 does not solve the issue. I also tried setting the logging level to debug but the output log is the exactly the same when I open any Platform.io workspace.
I saw that the version 1.2.1 was released today, perhaps I am the first reporting it.
The only extensions I have installed are Bracket Pair Colorizer 2, Jupyter, Pylance, Python, PlatformIO IDE and C/C++.

**Comment by Colengms** [comment]
Hi @gunarkroeger .  I have a simple fix in PR that would seem to address the symptom, but I can't figure out how the scenario you are encountering is possible.  Considering it doesn't always happen (I don't repro the issue, and our testing didn't encounter it), we could really use more details about your setup.  Is this repro'ing with a multiroot workspace?  (The code path is slightly different for single-root vs multiroot).  If so, does it also repro when opening just a single folder?  Does it repro when opening a workspace with no files, then opening a file?  Or, only when files are re-opened on launch?  (The error appears related to processing a didOpen message, before everything is set up).  Can you provide any `C_Cpp` settings you have changed from default?

**Comment by gunarkroeger** [comment]
Hi @Colengms . I'm sorry for not being able to do much more testing on this for you. I can say that I have tested with 2 separate platform.io single root workspaces with the same result. I quickly checked my C_Cpp settings and there it's all on the default value. My projects have clang auto-formating set up, not sure if this is relevant.
With this error the Platform.io icon doesn't even show up in the sidebar.

To get it working, I just roll back the version to 1.2.0 and reload the window.
I might be able to do more testing for you next week. Hopefully, someone else who encounters this problem can provide more details before then.

**Comment by steves2j** [comment]
Hi
Running VS on a mac, I'm also seeing this issue with this extension with the same exception. Rolling back to 1.2.0 fixes the issue. However, vscode keeps trying to update it back to 1.2.1 after rolling it back. Is there any way to force it not to update this extension?

**Comment by sean-mcmanus** [comment]
Can you two try https://github.com/microsoft/vscode-cpptools/releases/tag/1.2.2-insiders ? That has a potential fix, although I don't think we understand the root cause, so it's possible there could be other errors reported.

@steves2j You can set extensions.autoUpdate to false, but you should try 1.2.2-insiders first.

**Comment by steves2j** [comment]
@sean-mcmanus 1.2.2-insiders installs and does not error as before, however it breaks some other extensions that depend on it like Particles Workbench and Platform IO.

**Comment by sean-mcmanus** [comment]
What do you mean by "break"? How are those tools broken? Normally, extensions can't break other extensions. Is there an error when you do Help->Toggle Developer Tools->Console?

**Comment by steves2j** [comment]
Normally we would have a Particle Workbench Welcome panel shown and Commands to Particle: compile and Particle: Upload etc. However when loading 1.2.2 these all disappear from the command options and the Particle panel.
Loading the console shows the following.
`Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:1734
/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:1734 Cannot activate extension 'Workbench Core' because it depends on extension 'ms-vscode.cpptools', which failed to activate.
_logMessageInConsole @ /Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:1734
/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:1734 Cannot activate extension 'PlatformIO IDE' because it depends on extension 'ms-vscode.cpptools', which failed to activate.`

**Comment by sean-mcmanus** [comment]
The `'ms-vscode.cpptools', which failed to activate` error implies our fix didn't work and there was another failure...but you're not seeing any failure message?

@Colengms Any ideas?

**Comment by steves2j** [comment]
Correct. There is a sparse amount of errors produced.

**Comment by sWski** [comment]
Hello, I also ran into this issue, Looks like the extension won't start if the workspace contains folder that does not exists. 

I have 3 different folders in my workspace (one of them is the `gc-tools2` folder) and a few days back I've removed entire C:\Dropbox directory that contained this project. Version 1.2.0 did not complain and started succesfully. Version 1.2.1 starts only after removing `gc-tools2` folder from my workspace.

<details>
<summary>Attached debug output</summary>
<p>

```
[2021-02-18 09:34:20.785] [renderer1] [error] [File Watcher (node.js)] Error: ENOENT: no such file or directory, stat 'c:\Dropbox\Geocaching\gc-tools2\.vscode'
[2021-02-18 09:34:20.785] [renderer1] [warning] Ignoring the error while validating workspace folder file:///c%3A/Dropbox/Geocaching/gc-tools2 - Unable to resolve non-existing file 'c:\Dropbox\Geocaching\gc-tools2'
[2021-02-18 09:34:23.269] [renderer1] [error] Cannot read property 'ActiveClient' of undefined: TypeError: Cannot read property 'ActiveClient' of undefined
    at Object.<anonymous> (c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:6986:33)
    at Generator.next (<anonymous>)
    at c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:6812:71
    at new Promise (<anonymous>)
    at module.exports.__awaiter (c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:6808:12)
    at Object.provideCodeActions (c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:6985:66)
    at c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:86:143067
    at c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:55:8183
    at new Promise (<anonymous>)
    at Object.S [as asPromise] (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:55:8155)
    at w.provideCodeActions (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:86:143038)
    at c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:86:165342
    at we._withAdapter (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:86:159898)
    at we.$provideCodeActions (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:86:165318)
    at p._doInvokeHandler (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:89:10276)
    at p._invokeHandler (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:89:9968)
    at p._receiveRequest (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:89:8587)
    at p._receiveOneMessage (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:89:7440)
    at c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:89:5568
    at v.fire (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:55:1836)
    at d.fire (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:63:15515)
    at c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:104:29104
    at v.fire (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:55:1836)
    at d.fire (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:63:15515)
    at t._receiveMessage (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:63:20765)
    at c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:63:17659
    at v.fire (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:55:1836)
    at u.acceptChunk (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:63:12880)
    at c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:63:12228
    at Socket.m (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:104:12375)
    at Socket.emit (events.js:315:20)
    at addChunk (_stream_readable.js:295:12)
    at readableAddChunk (_stream_readable.js:271:9)
    at Socket.Readable.push (_stream_readable.js:212:10)
    at Pipe.onStreamRead (internal/stream_base_commons.js:186:23)
[2021-02-18 09:34:24.446] [renderer1] [error] ENOENT: no such file or directory, scandir 'c:\Dropbox\Geocaching\gc-tools2': Error: ENOENT: no such file or directory, scandir 'c:\Dropbox\Geocaching\gc-tools2'
[2021-02-18 09:34:25.834] [renderer1] [error] Attempting to use languageClient before initialized: Error: Attempting to use languageClient before initialized
    at DefaultClient.get languageClient [as languageClient] (c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:3928:19)
    at c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:4331:18
    at c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:4843:21
    at new Promise (<anonymous>)
    at task (c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:4842:28)
    at DefaultClient.<anonymous> (c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:4791:34)
    at Generator.next (<anonymous>)
    at c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:3458:71
    at new Promise (<anonymous>)
    at module.exports.__awaiter (c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:3454:12)
    at nextTask (c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:4789:36)
[2021-02-18 09:34:29.421] [renderer1] [error] Cannot read property 'ActiveClient' of undefined: TypeError: Cannot read property 'ActiveClient' of undefined
    at Object.<anonymous> (c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:6986:33)
    at Generator.next (<anonymous>)
    at c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:6812:71
    at new Promise (<anonymous>)
    at module.exports.__awaiter (c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:6808:12)
    at Object.provideCodeActions (c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:6985:66)
    at c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:86:143067
    at c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:55:8183
    at new Promise (<anonymous>)
    at Object.S [as asPromise] (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:55:8155)
    at w.provideCodeActions (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:86:143038)
    at c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:86:165342
    at we._withAdapter (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:86:159898)
    at we.$provideCodeActions (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:86:165318)
    at p._doInvokeHandler (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:89:10276)
    at p._invokeHandler (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:89:9968)
    at p._receiveRequest (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:89:8587)
    at p._receiveOneMessage (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:89:7440)
    at c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:89:5568
    at v.fire (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:55:1836)
    at d.fire (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:63:15515)
    at c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:104:29104
    at v.fire (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:55:1836)
    at d.fire (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:63:15515)
    at t._receiveMessage (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:63:20765)
    at c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:63:17659
    at v.fire (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:55:1836)
    at u.acceptChunk (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:63:12880)
    at c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:63:12228
    at Socket.m (c:\Users\<user>\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:104:12375)
    at Socket.emit (events.js:315:20)
    at addChunk (_stream_readable.js:295:12)
    at readableAddChunk (_stream_readable.js:271:9)
    at Socket.Readable.push (_stream_readable.js:212:10)
    at Pipe.onStreamRead (internal/stream_base_commons.js:186:23)
```

</p>
</details>

<details>
<summary>Extension error output</summary>
<p>

```
[Error - 10:05:12] Server initialization failed.
TypeError: Cannot read property 'setDidOpenTime' of undefined
    at Object.didOpen (c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:60328:52)
    at c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:8696:32
    at Array.forEach (<anonymous>)
    at DidOpenTextDocumentFeature.register (c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:8685:42)
    at DidOpenTextDocumentFeature.initialize (c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:8676:18)
    at LanguageClient.initializeFeatures (c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:10684:21)
    at c:\Users\<user>\.vscode\extensions\ms-vscode.cpptools-1.2.1\dist\main.js:10389:18
```

</p>
</details>

**Comment by CJxD** [comment]
I also have this issue with cpptools:1.2.1 which has started since yesterday. Running VSCode 1.53.2 on Ubuntu 20.04. Unfortunately I'm not able to get the 1.2.2-insiders update to try.

**Comment by modbw** [comment]
I had the exact same problem starting with 1.2.1. I can confirm that after creating all paths in my workspace the problem disappears.

**Comment by mario-almeida** [comment]
Exact same problem with 1.2.1, issue resolved after removing invalid paths from workspace file.

**Comment by chrisckc** [comment]
Just found the same with 1.2.1 and then noticed that the workspace referenced a directory which I had renamed in the past and the update to 1.2.1 now makes it crash.

**Comment by sean-mcmanus** [comment]
Thanks for the repro info. We should have a fix soon for 1.2.2.

**Comment by sean-mcmanus** [comment]
Fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/1.2.2-insiders2
